payments/receipt_manager.go

```diff
@@ -0,0 +1,91 @@
+package payments
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/statechannels/go-nitro/types"
+)
+
+type (
+	// PaymentStatus stores the status of payments for a given payment channel.
+	PaymentStatus struct {
+		channelSender   common.Address
+		startingBalance *big.Int
+		largestVoucher  Voucher
+	}
+
+	// ReceiptManager receives vouchers, validates them, and stores the most valuable voucher
+	ReceiptManager struct {
+		channels map[types.Destination]*PaymentStatus
+	}
+)
+
+func NewReceiptManager() *ReceiptManager {
+	channels := make(map[types.Destination]*PaymentStatus)
+	return &ReceiptManager{channels}
+}
+
+// Register registers a channel for use, given the
+func (pm ReceiptManager) Register(channelId types.Destination, sender common.Address, startingBalance *big.Int) error {
+	balance := &big.Int{}
+	balance.Set(startingBalance)
+	voucher := Voucher{channelId: channelId, amount: big.NewInt(0)}
+	data := &PaymentStatus{sender, balance, voucher}
+	if _, ok := pm.channels[channelId]; ok {
+		return fmt.Errorf("channel already registered")
+	}
+
+	pm.channels[channelId] = data
+
+	return nil
+}
+
+// Remove deletes the channel's status
+func (pm *ReceiptManager) Remove(channelId types.Destination) {
+	delete(pm.channels, channelId)
+}
+
+// Receive validates the incoming voucher, and returns the total amount received so far
+func (rm *ReceiptManager) Receive(voucher Voucher) (*big.Int, error) {
+	status, ok := rm.channels[voucher.channelId]
+	if !ok {
+		return &big.Int{}, fmt.Errorf("channel not registered")
```

Comment:
What if the channel was actually not yet / np longer open or not yet / no longer funded? Shouldn't we return an error in that case?

Reply:
Would all those cases be covered by "this channel is not registered"?

My mental model is: a channel is either in a state where it can receive vouchers, or it is not in that state. It might not be in that state for many reasons (it's a made up channel, it's being de-funded, etc.). But the receipt manager doesn't care about that -- it only cares about managing vouchers.

Reply:
> Would all those cases be covered by "this channel is not registered"?

I think in principle, yes. I'm just wondering about keeping the register of channels in sync with the client's store. But your response to my other question addresses that.